sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala

```diff
@@ -84,19 +84,46 @@ object ReorderJoin extends Rule[LogicalPlan] with PredicateHelper {
     }
   }
 
+  private def extractLeftDeepInnerJoins(plan: LogicalPlan): Seq[LogicalPlan] = plan match {
```

Comment:
It took me a little while to understand that this can handle `(a join b) join c` versus `a join (b join c)` correctly. Would be great if we can explain how it works in the function comment.